ensure that regional inductor uses a boxed calling convention

bdhirsh opened the pull request:
I noticed when running regional inductor that it would emit the unboxed compiler warnings:
```
UserWarning: Your compiler for AOTAutograd is returning a function that doesn't take boxed arguments
```

I tried tweaking the regional_inductor backend to use boxed codegen. Two followup changes I needed were:
(1) making it pickleable
(2) Ensuring that `invoke_subgraph()` knows how to handle the case where the subgraph it calls expects boxed inputs

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #169295
* #169023
* __->__ #168277



cc @ezyang @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @chenyang78

bdhirsh committed:
ensure that regional inductor uses a boxed calling convention

[ghstack-poisoned]

anijain2305 reviewed (approved):
Thank you!

ezyang commented on torch/fx/passes/regional_inductor.py:
```diff
@@ -141,7 +172,13 @@ def is_node_supported(self, submodules, node: torch.fx.Node) -> bool:
+    gm.recompile()
+
+    # Use a pickleable wrapper class that provides boxed calling convention
+    wrapper = _BoxedCallWrapper(gm)
```
hang on, why? Once you set codegen on gm, it should have boxed calling convention no?

ezyang reviewed (approved):
Approving to unblock but please check comments

bdhirsh committed:
Update on "ensure that regional inductor uses a boxed calling convention"

I noticed when running regional inductor that it would emit the unboxed compiler warnings:
```
UserWarning: Your compiler for AOTAutograd is returning a function that doesn't take boxed arguments
```

I tried tweaking the regional_inductor backend to use boxed codegen. Two followup changes I needed were:
(1) making it pickleable
(2) Ensuring that `invoke_subgraph()` knows how to handle the case where the subgraph it calls expects boxed inputs




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

bdhirsh committed:
Update on "ensure that regional inductor uses a boxed calling convention"

I noticed when running regional inductor that it would emit the unboxed compiler warnings:
```
UserWarning: Your compiler for AOTAutograd is returning a function that doesn't take boxed arguments
```

I tried tweaking the regional_inductor backend to use boxed codegen. Two followup changes I needed were:
(1) making it pickleable
(2) Ensuring that `invoke_subgraph()` knows how to handle the case where the subgraph it calls expects boxed inputs




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

bdhirsh committed:
Update on "ensure that regional inductor uses a boxed calling convention"

I noticed when running regional inductor that it would emit the unboxed compiler warnings:
```
UserWarning: Your compiler for AOTAutograd is returning a function that doesn't take boxed arguments
```

I tried tweaking the regional_inductor backend to use boxed codegen. Two followup changes I needed were:
(1) making it pickleable
(2) Ensuring that `invoke_subgraph()` knows how to handle the case where the subgraph it calls expects boxed inputs




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

bdhirsh committed:
Update on "ensure that regional inductor uses a boxed calling convention"

I noticed when running regional inductor that it would emit the unboxed compiler warnings:
```
UserWarning: Your compiler for AOTAutograd is returning a function that doesn't take boxed arguments
```

I tried tweaking the regional_inductor backend to use boxed codegen. Two followup changes I needed were:
(1) making it pickleable
(2) Ensuring that `invoke_subgraph()` knows how to handle the case where the subgraph it calls expects boxed inputs




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

bdhirsh committed:
Update on "ensure that regional inductor uses a boxed calling convention"

I noticed when running regional inductor that it would emit the unboxed compiler warnings:
```
UserWarning: Your compiler for AOTAutograd is returning a function that doesn't take boxed arguments
```

I tried tweaking the regional_inductor backend to use boxed codegen. Two followup changes I needed were:
(1) making it pickleable
(2) Ensuring that `invoke_subgraph()` knows how to handle the case where the subgraph it calls expects boxed inputs




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

bdhirsh committed:
Update on "ensure that regional inductor uses a boxed calling convention"

I noticed when running regional inductor that it would emit the unboxed compiler warnings:
```
UserWarning: Your compiler for AOTAutograd is returning a function that doesn't take boxed arguments
```

I tried tweaking the regional_inductor backend to use boxed codegen. Two followup changes I needed were:
(1) making it pickleable
(2) Ensuring that `invoke_subgraph()` knows how to handle the case where the subgraph it calls expects boxed inputs




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

bdhirsh committed:
Update on "ensure that regional inductor uses a boxed calling convention"

I noticed when running regional inductor that it would emit the unboxed compiler warnings:
```
UserWarning: Your compiler for AOTAutograd is returning a function that doesn't take boxed arguments
```

I tried tweaking the regional_inductor backend to use boxed codegen. Two followup changes I needed were:
(1) making it pickleable
(2) Ensuring that `invoke_subgraph()` knows how to handle the case where the subgraph it calls expects boxed inputs




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge
